### PR TITLE
feat: `avm/res/service-fabric/cluster`

### PR DESCRIPTION
**Pull request by snoejovich**
## Description

Migrate bicep module for service-fabric cluster from CARML.

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.service-fabric.cluster](https://github.com/snoejovich/bicep-registry-modules/actions/workflows/avm.res.service-fabric.cluster.yml)    |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: .github/CODEOWNERS
Comment (AlexanderSehr) [inline .github/CODEOWNERS:129]: Soft-block for this file until the GitHub teams are confirmed in the hierachy

File: avm/res/service-fabric/cluster/tests/e2e/cert/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/service-fabric/cluster/tests/e2e/cert/main.test.bicep:72]:
```suggestion
```

The general guidance for the test files would be to try and keep them 'on point'. This somewhat cryptic statement just means that we want to ensure the test cases show a specific scenario and do no introduce unrelated test values. In this case, for example, we''d not want to add 'tags' to the test (which are already tested via the max test) because it would give the impression in the readme's Usage Examples that a user would need to provide tags in order to deploy the module with a certificate. I hope that makes somewhat sense. 👀

File: avm/res/service-fabric/cluster/tests/e2e/max/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/service-fabric/cluster/tests/e2e/max/main.test.bicep:230]: Quick question on this one - if the value is not idempotent, shouldn't it then maybe be moved into a dedicated test that does not perform an indempotency test? I'm just curious as this deployment here runs in a loop that is executed twice - so either it doesn't fail (which makes it idempotent) and the comment is wrong, or something is off :D Your pipeline run log for this test looks good.

File: avm/res/service-fabric/cluster/tests/e2e/waf-aligned/main.test.bicep
Comment (AlexanderSehr) [inline avm/res/service-fabric/cluster/tests/e2e/waf-aligned/main.test.bicep:58]:
```suggestion
```

not required for WAF alignment. Same as with the `cert` test the suggestion would be to question the parameters that are provided and only add those that are required for waf alignment :)

Comment (AlexanderSehr) [reply avm/res/service-fabric/cluster/tests/e2e/waf-aligned/main.test.bicep:58]: Note: If you're unsure, I recommend to remove all parameters but the minimum required, run the pipeline once and check if PSRule complains about anything in the pipeline.

**Comment by AlexanderSehr** [comment]
Hey @snoejovich, thank you for the great contribution. Not much to note, just added a few comments/suggestions 💪

File: avm/res/service-fabric/cluster/tests/e2e/max/main.test.bicep
Comment (snoejovich) [reply avm/res/service-fabric/cluster/tests/e2e/max/main.test.bicep]: Agreed that the test passed and we saw that this comment was in the original CARML repo, so we think it can just be removed.

**Comment by oZakari** [comment]
@AlexanderSehr for the 'validate-non-module-files' the file it is referring to is not a file that is modified within this PR. Is this something that we need to modify for this PR?

cc: @snoejovich 

image

**Comment by AlexanderSehr** [comment]
> @AlexanderSehr for the 'validate-non-module-files' the file it is referring to is not a file that is modified within this PR. Is this something that we need to modify for this PR?
> 
> cc: @snoejovich
> 
> image

Hey @oZakari, a curious fact indeed given that the file wasn't touched in the PR. Tried to resolve it by re-openining the PR, but the issue just returned. We wanted to deprecate prettier anyways, so we may just need to do this sooner rather than later. Very annoying :/ 

The PR  otherwise looks good @snoejovich. I'd just ask you to re-run the `Set-AVMModule` script to ensure all latest changes are applied :)

File: avm/res/service-fabric/cluster/application-type/main.bicep
Comment (eriqua) [inline avm/res/service-fabric/cluster/application-type/main.bicep:1]: Hey @snoejovich, the latest bicep version introduced the enforcement of v2 formatting by default for all bicep files. 
Thus, we'd kindly require the following changes to enforce v2 formatting for this module:
- merge latest upstream main to this branch fork
- format all bicep files of this module to comply with v2. This includes main.bicep but also all main.test.bicep and dependencies.bicep files. To do that it's enough to add a space anywhere in the file and save. The format on save repo setting will automatically format the file using v2

Comment (snoejovich) [reply avm/res/service-fabric/cluster/application-type/main.bicep:1]: Thank you, we have made the necessary changes.

**Comment by lsnoddy** [comment]
@snoejovich,  It looks like a copy of the service-fabric/cluster folder was inadvertently included directly beneath the avm folder. Please remove that folder from your featur branch and push the changes.

image

File: avm/res/service-fabric/cluster/main.bicep
Comment (AlexanderSehr) [inline avm/res/service-fabric/cluster/main.bicep:211]: It's a minor thing / suggestion so feel free to just resolve it, but if you want to simply these statements you can use the tryGet assert (`?`) in combination with the coalesce (`??`)

For example, instead of 
```bicep
maxPercentDeltaUnhealthyNodes: contains(upgradeDescription, 'maxPercentDeltaUnhealthyNodes')
        ? upgradeDescription.maxPercentDeltaUnhealthyNodes
        : 0
```

you could simply it to
```bicep
maxPercentDeltaUnhealthyNodes: upgradeDescription.?maxPercentDeltaUnhealthyNodes ?? 0
```

Comment (AlexanderSehr) [reply avm/res/service-fabric/cluster/main.bicep:211]: Again, no need to 'actually do it'. Just figured I should mention it as the formatter took a swing at it 😄

**Review by AlexanderSehr — DISMISSED** [review]

**Review by AlexanderSehr — APPROVED** [review]

File: .github/CODEOWNERS
Comment (AlexanderSehr) [reply .github/CODEOWNERS:129]: Hey @snoejovich, from where I'm standing the PR looks ready to be merged, but the team is either not yet created, or not yet added in the correct location in the hierachy so that the CODEOWNERS file will work as expected (hence it shows errors in the `Files changed` view of the Pull Request. You can find more information on what exactly to do in [this](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr20---category-contributionsupport---github-teams-only) spec.

Comment (lsnoddy) [reply .github/CODEOWNERS:129]: The parent team assignment was missing. I've added the avm-technical-reviewers-bicep parent team and am awaiting  the approval.

**Comment by lsnoddy** [comment]
Completes https://github.com/Azure/Azure-Verified-Modules/issues/562

**Pull request merged**
